File: java/src/main/java/io/misar/mail/core/Sse.java
package io.misar.mail.core;

import io.misar.mail.MisarMailException;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.function.Consumer;

/**
 * Server-Sent Events client for the MisarMail streaming endpoints.
 *
 * <p>Both streams frame events as {@code data: <json>} and close with the
 * sentinel {@code data: [DONE]}. One of the two is a POST, so the stream is
 * driven off a normal request with a streamed body rather than an
 * EventSource-style helper.
 */
public final class Sse {

  private static final String DONE = "[DONE]";

  private Sse() {}

  /**
   * Opens an SSE endpoint and hands each decoded frame to {@code onEvent} until
   * the stream terminates. Blocks the calling thread for the life of the stream.
   */
  public static void stream(
      String url, String apiKey, String method, String jsonBody, Consumer<Map<String, Object>> onEvent)
      throws MisarMailException {

    // A stream has no useful deadline, so it deliberately gets its own client
    // rather than inheriting the transport's per-request timeout.
    HttpClient http = HttpClient.newHttpClient();

    HttpRequest.Builder builder =
        HttpRequest.newBuilder(URI.create(url))
            .header("Authorization", "Bearer " + apiKey)
            .header("Accept", "text/event-stream");

    if (jsonBody != null) {
      builder.header("Content-Type", "application/json");
      builder.method(method, HttpRequest.BodyPublishers.ofString(jsonBody));
    } else {
      builder.method(method, HttpRequest.BodyPublishers.noBody());
    }

    try {
      HttpResponse<InputStream> response =
          http.send(builder.build(), HttpResponse.BodyHandlers.ofInputStream());

      if (response.statusCode() >= 400) {
        // Errors arrive as a normal JSON body, not as an SSE frame.
        String raw = new String(response.body().readAllBytes(), StandardCharsets.UTF_8);
        Map<String, Object> data = Json.readObject(raw);
        throw new MisarMailException(
            response.statusCode(),
            String.valueOf(data.getOrDefault("error", "stream failed")),
            "api_error");
      }

      try (BufferedReader reader =
          new BufferedReader(new InputStreamReader(response.body(), StandardCharsets.UTF_8))) {
        String line;
        while ((line = reader.readLine()) != null) {
          if (!line.startsWith("data:")) {
            continue;
          }
          String payload = line.substring(5).trim();
          if (DONE.equals(payload)) {
            return;
          }
          if (payload.isEmpty()) {
            continue;
          }
          Map<String, Object> decoded = Json.readObject(payload);
          // An unparseable frame surfaces as {"raw": ...} rather than killing
          // the stream and discarding everything already delivered.
          onEvent.accept(decoded.isEmpty() ? Map.of("raw", payload) : decoded);
        }
      }
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new MisarMailException(0, "interrupted", "network_error");
    } catch (MisarMailException e) {
      throw e;
    } catch (Exception e) {
      throw new MisarMailException(0, e.getMessage(), "network_error");
    }
  }
}

File: kotlin/src/main/kotlin/io/misar/mail/core/Sse.kt
package io.misar.mail.core

import io.misar.mail.MisarMailException
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn

/**
 * Server-Sent Events client for the MisarMail streaming endpoints.
 *
 * Both streams frame events as `data: <json>` and close with the sentinel
 * `data: [DONE]`. One of the two is a POST, so the stream is driven off a
 * normal request with a streamed body rather than an EventSource-style helper.
 */
object Sse {

    private const val DONE = "[DONE]"

    /**
     * Streams a MisarMail SSE endpoint as a cold [Flow], emitting one decoded
     * payload per event. Collecting the flow opens the connection; cancelling
     * the collector closes it.
     */
    fun stream(
        url: String,
        apiKey: String,
        method: String = "GET",
        body: Any? = null,
    ): Flow<Map<String, Any?>> = flow {
        // A stream has no useful deadline, so it gets its own client rather
        // than inheriting the transport's per-request timeout.
        val http = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(30)).build()

        val builder = HttpRequest.newBuilder(URI.create(url))
            .header("Authorization", "Bearer $apiKey")
            .header("Accept", "text/event-stream")

        if (body != null) {
            builder.header("Content-Type", "application/json")
            builder.method(method, HttpRequest.BodyPublishers.ofString(Json.write(body)))
        } else {
            builder.method(method, HttpRequest.BodyPublishers.noBody())
        }

        val response = http.send(builder.build(), HttpResponse.BodyHandlers.ofLines())

        if (response.statusCode() >= 400) {
            // Errors arrive as a normal JSON body, not as an SSE frame.
            val raw = response.body().toList().joinToString("\n")
            val data = Json.readObject(raw)
            throw MisarMailException(
                response.statusCode(),
                (data["error"] ?: "stream failed").toString(),
                "api_error",
            )
        }

        for (line in response.body()) {
            if (!line.startsWith("data:")) continue

            val payload = line.substring(5).trim()
            if (payload == DONE) return@flow
            if (payload.isEmpty()) continue

            val decoded = Json.readObject(payload)
            // One malformed frame should not discard everything already streamed.
            emit(if (decoded.isEmpty()) mapOf("raw" to payload) else decoded)
        }
    }.flowOn(Dispatchers.IO)
}
